### PR TITLE
Add admin API to reset connection timeouts for remote server

### LINKED ISSUE #1266: We need an admin API to reset federation retries for a given HS (SYN-222)
If as an admin I know that a destination HS has come back online, I need way of updating the retry DB to tell my one to retry asap rather than waiting for the next timeout.

(Imported from https://matrix.org/jira/browse/SYN-222)

(Reported by @ara4n)

### PR DESCRIPTION
**Pull request by dklimpel**
Fixes #1266
Based on #11593

Add admin API
```
POST /_synapse/admin/v1/federation/destinations/<destination>/reset_connection
{}
```

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: synapse/rest/admin/federation.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 from typing import TYPE_CHECKING, Tuple
 
 from synapse.api.errors import Codes, NotFoundError, SynapseError
+from synapse.federation.transport.server._base import Authenticator
```
Comment (dklimpel) [inline synapse/rest/admin/federation.py:19]: I got the error
```
builtins.ImportError: cannot import name 'Authenticator' from partially initialized module 'synapse.federation.transport.server._base' (most likely due to a circular import) (/synapse/build/synapse/federation/transport/server/_base.py)
```

I do not know how is the best way to solve this.
One option is do not use `_reset_retry_timings`
https://github.com/matrix-org/synapse/blob/66d7aa783a8ad1b3eb6d17ba086e2dda8a912ab6/synapse/federation/transport/server/_base.py#L121-L137

and call manually
```python
store.set_destination_retry_timings(origin, None, 0, 0)
notifier.notify_remote_server_up(origin)
```

Comment (DMRobertson) [reply synapse/rest/admin/federation.py]: Looks like you got this working, but it's a bit of a shame to have to duplicate those two lines. Was there any additional context with the import error?

Comment (DMRobertson) [reply synapse/rest/admin/federation.py]: Aha:

```
2021-12-23T16:24:01.3115062Z Traceback (most recent call last):
2021-12-23T16:24:01.3117986Z   File "/home/runner/work/synapse/synapse/.tox/py-noextras/lib/python3.10/site-packages/twisted/trial/runner.py", line 693, in loadByName
2021-12-23T16:24:01.3119093Z     return self.suiteFactory([self.findByName(name, recurse=recurse)])
2021-12-23T16:24:01.3120526Z   File "/home/runner/work/synapse/synapse/.tox/py-noextras/lib/python3.10/site-packages/twisted/trial/runner.py", line 453, in findByName
2021-12-23T16:24:01.3121458Z     obj = reflect.namedModule(searchName)
2021-12-23T16:24:01.3122925Z   File "/home/runner/work/synapse/synapse/.tox/py-noextras/lib/python3.10/site-packages/twisted/python/reflect.py", line 157, in namedModule
2021-12-23T16:24:01.3123764Z     topLevel = __import__(name)
2021-12-23T16:24:01.3124407Z   File "/home/runner/work/synapse/synapse/tests/api/test_auth.py", line 32, in <module>
2021-12-23T16:24:01.3125015Z     from tests import unittest
2021-12-23T16:24:01.3125657Z   File "/home/runner/work/synapse/synapse/tests/unittest.py", line 53, in <module>
2021-12-23T16:24:01.3126542Z     from synapse.federation.transport import server as federation_server
2021-12-23T16:24:01.3127524Z   File "/home/runner/work/synapse/synapse/synapse/federation/transport/server/__init__.py", line 21, in <module>
2021-12-23T16:24:01.3128527Z     from synapse.federation.transport.server._base import (
2021-12-23T16:24:01.3129664Z   File "/home/runner/work/synapse/synapse/synapse/federation/transport/server/_base.py", line 32, in <module>
2021-12-23T16:24:01.3130439Z     from synapse.server import HomeServer
2021-12-23T16:24:01.3131161Z   File "/home/runner/work/synapse/synapse/synapse/server.py", line 52, in <module>
2021-12-23T16:24:01.3131941Z     from synapse.events.spamcheck import SpamChecker
2021-12-23T16:24:01.3132802Z   File "/home/runner/work/synapse/synapse/synapse/events/spamcheck.py", line 31, in <module>
2021-12-23T16:24:01.3133605Z     from synapse.rest.media.v1._base import FileInfo
2021-12-23T16:24:01.3134368Z   File "/home/runner/work/synapse/synapse/synapse/rest/__init__.py", line 18, in <module>
2021-12-23T16:24:01.3255435Z     from synapse.rest import admin
2021-12-23T16:24:01.3256406Z   File "/home/runner/work/synapse/synapse/synapse/rest/admin/__init__.py", line 43, in <module>
2021-12-23T16:24:01.3257226Z     from synapse.rest.admin.federation import (
2021-12-23T16:24:01.3258116Z   File "/home/runner/work/synapse/synapse/synapse/rest/admin/federation.py", line 19, in <module>
2021-12-23T16:24:01.3259192Z     from synapse.federation.transport.server._base import Authenticator
2021-12-23T16:24:01.3261751Z builtins.ImportError: cannot import name 'Authenticator' from partially initialized module 'synapse.federation.transport.server._base' (most likely due to a circular import) 
```

To import `_base`, we have to import `Homeserver`. To do that, we have to import `synapse.rest`. Which with this change, imports `Authenticator` from `_base`: a loop.

We could hack around this by e.g. only importing `Homeserver` in `_base` if `typing.TYPE_CHECKING` and changing the type annotations to be strings. Not convinced that's the right way to do it---would be interested in thoughts from @matrix-org/synapse-core here

Comment (richvdh) [reply synapse/rest/admin/federation.py]: it looks like @dklimpel has worked around this anyway now, but:

1. The `_` in `_base` denotes it as a "private" module -  ie, it is only for use within `synapse.federation.transport.server, so it should not be imported into `synapse.rest.admin` anyway. (If you need to reference `Authenticator`, you can `from synapse.federation.transport.server import Authenticator`, but you'll still have a loop here).
2. Yes, `Homeserver` should probably only be imported if `typing.TYPE_CHECKING` - that's what most of the codebase does, for much the same reason.

`synapse.rest` is pretty high-level, so I think importing from `synapse.federation.transport.server` is probably ok.

File: docs/usage/administration/admin_api/federation.md [hunk 2]
```diff
@@ -108,7 +108,37 @@ A response body like the following is returned:
...
 
 The response fields are the same like in the `destinations` array in
 [List of destinations](#list-of-destinations) response.
+
+## Reset connection timeout
+
+This API resets the retry timing for a specific remote server and tries to connect
+the remote server again. It does not wait for the next `retry_interval`.
+The connection must have previously run into an error and `retry_last_ts`
+([Destination Details API](#destination-details-api)) must not be equal to `0`.
```
Comment (DMRobertson) [inline docs/usage/administration/admin_api/federation.md:127]: I think a description of synapse's retry behaviour would be helpful context before describing the API. Maybe something like this?

Synapse makes federation requests to other homeservers. If a federation request fails, Synapse will marks the destination homeserver as offline, preventing any future requests to that server for a "cooldown" period. This period grows over time if the server continues to fail its responses ([exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff)).

Admins can cancel the cooldown period with this API.

File: docs/usage/administration/admin_api/federation.md [hunk 3]
```diff
@@ -108,7 +108,37 @@ A response body like the following is returned:
...
+
 **Response**
 
 The response fields are the same like in the `destinations` array in
 [List of destinations](#list-of-destinations) response.
+
+## Reset connection timeout
+
+This API resets the retry timing for a specific remote server and tries to connect
+the remote server again. It does not wait for the next `retry_interval`.
```
Comment (DMRobertson) [inline docs/usage/administration/admin_api/federation.md:125]: > and tries to connect
> the remote server again

I think this is wrong. AFAICS writing `(None, 0, 0)` to the DB will allow future federation requests to be attempted, but there's no generic "queue" of requests that we've prevented due to previous requests failing.

**Review by DMRobertson — CHANGES_REQUESTED** [review]

File: synapse/rest/admin/federation.py [hunk 4]
```diff
@@ -145,3 +146,44 @@ async def on_GET(
...
+
+    def __init__(self, hs: "HomeServer"):
+        self._auth = hs.get_auth()
+        self._store = hs.get_datastore()
+        self._authenticator = Authenticator(hs)
+
+    async def on_POST(
+        self, request: SynapseRequest, destination: str
+    ) -> Tuple[int, JsonDict]:
+        await assert_requester_is_admin(self._auth, request)
```
Comment (DMRobertson) [inline synapse/rest/admin/federation.py:174]: Do we have to make this check explicitly for every admin API endpoint? I wonder if there's scope to introduce e.g. an `AdminRestServlet` class which does this automatically.

Comment (DMRobertson) [reply synapse/rest/admin/federation.py:174]: Oh ignore this, I said this on #11658.

File: docs/usage/administration/admin_api/federation.md (hunk 3 above)
Comment (DMRobertson) [reply docs/usage/administration/admin_api/federation.md:133]: Ahh, you're quite right. Thank you for explaining, and pointing out `notify_remote_server_up` in particular.

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Many thanks. I spotted a few typos (probably me), and there's the import that Rich pointed out. I think with those fixed we should be good to go!

**Review by DMRobertson — APPROVED** [review]
Many thanks; sorry this one took a bit longer than usual to review!

**Comment by dklimpel** [comment]
> Many thanks; sorry this one took a bit longer than usual to review!

Thanks. It is ok.

**Pull request merged**